lib/datadog/statsd/message_buffer.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module Datadog
+  class Statsd
+    class MessageBuffer
+      PAYLOAD_SIZE_TOLERANCE = 0.05
+
+      def initialize(connection,
+        max_buffer_payload_size: nil,
+        max_buffer_pool_size: DEFAULT_BUFFER_POOL_SIZE,
+        buffer_overflowing_stategy: :drop
+      )
+        raise ArgumentError, 'max_buffer_payload_size keyword argument must be provided' unless max_buffer_payload_size
+        raise ArgumentError, 'max_buffer_pool_size keyword argument must be provided' unless max_buffer_pool_size
+
+        @connection = connection
+        @max_buffer_payload_size = max_buffer_payload_size
+        @max_buffer_pool_size = max_buffer_pool_size
+        @buffer_overflowing_stategy = buffer_overflowing_stategy
+
+        @buffer = String.new
+        @message_count = 0
+
+        @depth = 0
+      end
+
+      def add(message)
+        message_size = message.bytesize
+
```

Comment:
I'm wondering if we should add a check on the message_size == 0, to avoid adding useless `\n` to the buffer.

Reply:
I think it's a good idea.